File: russian-doll-envelopes/russian-doll-envelopes.java
class Solution {
    public int maxEnvelopes(int[][] envelopes) {
        Arrays.sort(envelopes, new Comparator<int[]>(){
            public int compare(int[] a, int[] b){
                if(a[0] == b[0]) return b[1] - a[1];
                else return a[0] - b[0];
            }
        });

        int[] dp = new int[envelopes.length];
        for(int i = 0; i < dp.length; i++) dp[i] = envelopes[i][1];
        return longestIncreasingSubsequence(dp);
    }

    private int longestIncreasingSubsequence(int[] nums){
        List<Integer> subArr = new ArrayList<>();
        subArr.add(nums[0]);

        for(int i = 1; i < nums.length; i++){
            if(nums[i] > subArr.get(subArr.size() - 1)) subArr.add(nums[i]);
            else subArr.set(binarySearch(subArr, nums[i]), nums[i]);            
        }
        return subArr.size();
    }

    private int binarySearch(List<Integer> nums, int target){
        int left = 0, right = nums.size() - 1;
        while(left <= right){
            int mid = left + (right - left)/2;
            if(nums.get(mid) < target) left = mid + 1;
            else right = mid - 1;
        }
        return left;
    }
}

File: russian-doll-envelopes/russian-doll-envelopes.go
import "sort"

func binarySearch(nums []int, target int) int {
    left, right := 0, len(nums) - 1
    for left <= right {
        mid := left + (right - left)/2
        if nums[mid] < target {
            left = mid + 1
        }else{
            right = mid - 1
        }
    }
    return left
}

func LIS(nums []int) int{
    sub := []int{}
    sub = append(sub, nums[0])

    for i := 1; i < len(nums); i++ {
        if nums[i] > sub[len(sub) - 1] {
            sub = append(sub, nums[i])
        }else{
            sub[binarySearch(sub, nums[i])] = nums[i]
        }
    }
    return len(sub)
}
func maxEnvelopes(envelopes [][]int) int {
    sort.SliceStable(envelopes, func(i, j int) bool {
        if envelopes[i][0] == envelopes[j][0]{
            return envelopes[j][1] < envelopes[i][1]
        }else{
            return envelopes[i][0] < envelopes[j][0]
        }
    })

    dp := []int{}
    for i := 0; i < len(envelopes); i++ {
        dp = append(dp, envelopes[i][1])
    }    
    return LIS(dp)
}